### PR TITLE
explain: add aggregation

### PR DESCRIPTION
**Pull request by Joyinqin**
### What is changed, added or deleted? (Required)
This is a followup PR to #3858 and the content here is based on/duplicates query-execution-plan.

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs/pull/3926
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Have version specific changes 
- [ ] Might cause conflicts

### CODE REVIEW

File: explain-aggregation.md [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
++-----------+--------------+--------------+-----------+-----------------+-------+------------+---------------+------------+----------------------+------------------+
+|        64 | t_64_        | t_64_r_31766 |        65 |               1 | 65    |          0 |          1325 |  102033520 |                   98 |            52797 |
+|        66 | t_64_r_31766 | t_64_r_63531 |        67 |               1 | 67    |          0 |          1325 |   72522521 |                  104 |            78495 |
+|        68 | t_64_r_63531 | t_64_r_95296 |        69 |               1 | 69    |          0 |          1325 |          0 |                  104 |            95433 |
+|         2 | t_64_r_95296 |              |         3 |               1 | 3     |          0 |          1501 |          0 |                   81 |            63211 |
++-----------+--------------+--------------+-----------+-----------------+-------+------------+---------------+------------+----------------------+------------------+
+4 rows in set (0.00 sec)
+```
+
+`EXPLAIN` 语句与以下的聚合语句一起使用，可以看到在 TiKV 内的每个 Region 上先执行 `└─StreamAgg_8`；然后，每个 TiKV Region 会发送一行返回 TiDB，该行将聚合来自 `StreamAgg_16` 中每个 Region 的数据：
```
Comment (TomShawn) [inline explain-aggregation.md:59]:
```suggestion
使用 `EXPLAIN` 查看以下聚合语句的执行计划。可以看到 `└─StreamAgg_8` 算子先执行在 TiKV 内每个 Region 上，然后 TiKV 的每个 Region 会返回一行数据给 TiDB，TiDB 在 `StreamAgg_16` 算子上对每个 Region 返回的数据进行聚合：
```

File: explain-aggregation.md [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
++----------------------------+-----------+-----------+---------------+---------------------------------+
+| StreamAgg_16               | 1.00      | root      |               | funcs:count(Column#7)->Column#5 |
+| └─TableReader_17           | 1.00      | root      |               | data:StreamAgg_8                |
+|   └─StreamAgg_8            | 1.00      | cop[tikv] |               | funcs:count(1)->Column#7        |
+|     └─TableFullScan_15     | 242020.00 | cop[tikv] | table:t1      | keep order:false                |
++----------------------------+-----------+-----------+---------------+---------------------------------+
+4 rows in set (0.00 sec)
+```
+
+同样，执行 `EXPLAIN ANALYZE` 语句后可知，因为正在使用 `TableFullScan` 并且没有二级索引，`actRows` 与 `SHOW TABLE REGIONS` 中的 Region 数匹配：
```
Comment (TomShawn) [inline explain-aggregation.md:79]:
```suggestion
同样，通过执行 `EXPLAIN ANALYZE` 语句可知，`actRows` 与 `SHOW TABLE REGIONS` 返回结果中的 Region 数匹配，这是因为执行使用了 `TableFullScan` 全表扫并且没有二级索引：
```

File: explain-aggregation.md [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+|     └─TableFullScan_15     | 242020.00 | 121010  | cop[tikv] | table:t1      | proc max:12ms, min:12ms, p80:12ms, p95:12ms, iters:122, tasks:4                                                                                                                                                                                 | keep order:false                | N/A       | N/A  |
++----------------------------+-----------+---------+-----------+---------------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+---------------------------------+-----------+------+
+4 rows in set (0.01 sec)
+```
+
+## Hash Aggregation
+
+Hash Aggregation 算法在执行聚合时使用 Hash 表存储中间结果。此算法采用多线程并发优化，执行速度快，但与 Stream Aggregation 算法相比会消耗较多内存。
+
+下面是一个使用 Hash Aggregation 的例子：
```
Comment (TomShawn) [inline explain-aggregation.md:101]:
```suggestion
下面是一个使用 Hash Aggregation（即 `HashAgg` 算子）的例子：
```

File: explain-aggregation.md [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
++---------------------------+-----------+-----------+---------------+---------------------------------+
+| HashAgg_9                 | 1.00      | root      |               | funcs:count(Column#6)->Column#5 |
+| └─TableReader_10          | 1.00      | root      |               | data:HashAgg_5                  |
+|   └─HashAgg_5             | 1.00      | cop[tikv] |               | funcs:count(1)->Column#6        |
+|     └─TableFullScan_8     | 242020.00 | cop[tikv] | table:t1      | keep order:false                |
++---------------------------+-----------+-----------+---------------+---------------------------------+
+4 rows in set (0.00 sec)
+```
+
+`operator info` 列显示了用于聚合数据的 Hash 函数为 `funcs:count(1)->Column#6`。
```
Comment (TomShawn) [inline explain-aggregation.md:121]:
```suggestion
`operator info` 列显示，用于聚合数据的 Hash 函数为 `funcs:count(1)->Column#6`。
```

File: explain-aggregation.md [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+## Stream Aggregation
+
+Stream Aggregation 算法通常会比 Hash Aggregation 算法占用更少的内存。但是此算法要求按顺序发送数据，以便对依次到达的值实现流式数据聚合。
```
Comment (TomShawn) [inline explain-aggregation.md:125]:
```suggestion
Stream Aggregation 算法通常会比 Hash Aggregation 算法占用更少的内存。但是此算法要求数据按顺序发送，以便对依次到达的值实现流式数据聚合。
```

File: explain-aggregation.md [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+| Projection_4                 | 8000.00  | root      |               | test.t2.col1, Column#3                                                                      |
+| └─StreamAgg_8                | 8000.00  | root      |               | group by:test.t2.col1, funcs:count(1)->Column#3, funcs:firstrow(test.t2.col1)->test.t2.col1 |
+|   └─Sort_13                  | 10000.00 | root      |               | test.t2.col1                                                                                |
+|     └─TableReader_12         | 10000.00 | root      |               | data:TableFullScan_11                                                                       |
+|       └─TableFullScan_11     | 10000.00 | cop[tikv] | table:t2      | keep order:false, stats:pseudo                                                              |
++------------------------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------+
+5 rows in set (0.00 sec)
+```
+
+在上述示例中, 可以通过在 `col1` 上添加索引来消除 `└─Sort_13` 算子。 一旦添加了索引，就可以按顺序读取数据，并且可以消除 `└─Sort_13` 算子。
```
Comment (TomShawn) [inline explain-aggregation.md:153]:
```suggestion
以上示例中, 可以在 `col1` 上添加索引来消除 `└─Sort_13` 算子。添加索引后，TiDB 就可以按顺序读取数据并消除 `└─Sort_13` 算子。
```

File: explain-aggregation.md [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+| Projection_4                 | 8000.00  | root      |               | test.t2.col1, Column#3                                                                      |
+| └─StreamAgg_8                | 8000.00  | root      |               | group by:test.t2.col1, funcs:count(1)->Column#3, funcs:firstrow(test.t2.col1)->test.t2.col1 |
+|   └─Sort_13                  | 10000.00 | root      |               | test.t2.col1                                                                                |
+|     └─TableReader_12         | 10000.00 | root      |               | data:TableFullScan_11                                                                       |
+|       └─TableFullScan_11     | 10000.00 | cop[tikv] | table:t2      | keep order:false, stats:pseudo                                                              |
++------------------------------+----------+-----------+---------------+---------------------------------------------------------------------------------------------+
+5 rows in set (0.00 sec)
+```
+
+以上示例中, 可以在 `col1` 上添加索引来消除 `└─Sort_13` 算子。添加索引后，TiDB 就可以按顺序读取数据并消除 `└─Sort_13` 算子。
```
Comment (TomShawn) [inline explain-aggregation.md:153]:
```suggestion
以上示例中，可以在 `col1` 上添加索引来消除 `└─Sort_13` 算子。添加索引后，TiDB 就可以按顺序读取数据并消除 `└─Sort_13` 算子。
```

File: explain-aggregation.md [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+INSERT INTO t1 SELECT NULL, RANDOM_BYTES(1024), RANDOM_BYTES(1024), RANDOM_BYTES(1024) FROM t1 a JOIN t1 b JOIN t1 c LIMIT 10000;
+INSERT INTO t1 SELECT NULL, RANDOM_BYTES(1024), RANDOM_BYTES(1024), RANDOM_BYTES(1024) FROM t1 a JOIN t1 b JOIN t1 c LIMIT 10000;
+INSERT INTO t1 SELECT NULL, RANDOM_BYTES(1024), RANDOM_BYTES(1024), RANDOM_BYTES(1024) FROM t1 a JOIN t1 b JOIN t1 c LIMIT 10000;
+INSERT INTO t1 SELECT NULL, RANDOM_BYTES(1024), RANDOM_BYTES(1024), RANDOM_BYTES(1024) FROM t1 a JOIN t1 b JOIN t1 c LIMIT 10000;
+INSERT INTO t1 SELECT NULL, RANDOM_BYTES(1024), RANDOM_BYTES(1024), RANDOM_BYTES(1024) FROM t1 a JOIN t1 b JOIN t1 c LIMIT 10000;
+SELECT SLEEP(1);
+ANALYZE TABLE t1;
+```
+
+以上示例创建表格 `t1` 并插入数据后，再执行 [`SHOW TABLE REGIONS`](/sql-statements/sql-statement-show-table-regions.md) 语句。从以下输出结果可知，表 `t1` 已被切分为多个 Region：
```
Comment (TomShawn) [inline explain-aggregation.md:39]:
```suggestion
以上示例创建表格 `t1` 并插入数据后，再执行 [`SHOW TABLE REGIONS`](/sql-statements/sql-statement-show-table-regions.md) 语句。从以下 `SHOW TABLE REGIONS` 的执行结果可知，表 `t1` 被切分为多个 Region：
```

**Comment by TomShawn** [comment]
@lzmhhh123 Could you help take a look? Thanks!

**Review by lzmhhh123 — APPROVED** [review]
LGTM

File: explain-aggregation.md [hunk 9]
```diff
@@ -0,0 +1,174 @@
+---
+title: 用 EXPLAIN 查看聚合计算的 SQL 执行计划
```
Comment (TomShawn) [inline explain-aggregation.md:2]:
```suggestion
title: 用 EXPLAIN 查看带聚合计算的 SQL 执行计划
```

File: explain-aggregation.md [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+summary: 了解 TiDB 中 EXPLAIN 语句返回的执行计划信息。
+---
+
+# 用 EXPLAIN 查看聚合计算的 SQL 执行计划
```
Comment (TomShawn) [inline explain-aggregation.md:6]:
```suggestion
# 用 EXPLAIN 查看带聚合计算的 SQL 执行计划
```

File: explain-aggregation.md [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+
+SQL 优化器会选择以下任一算子实现数据聚合：
+
+- Hash Aggregation
+- Stream Aggregation
+
+为了提高查询效率，Coprocessor 和 TiDB 层均可执行数据聚合。现有示例如下：
```
Comment (TomShawn) [inline explain-aggregation.md:13]:
```suggestion
为了提高查询效率，数据聚合在 Coprocessor 和 TiDB 层均执行。现有示例如下：
```

File: explain-aggregation.md [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+```sql
+CREATE TABLE t2 (id INT NOT NULL PRIMARY KEY, col1 INT NOT NULL);
+INSERT INTO t2 VALUES (1, 9),(2, 3),(3,1),(4,8),(6,3);
+EXPLAIN SELECT /*+ STREAM_AGG() */ col1, count(*) FROM t2 GROUP BY col1;
+```
+
+```sql
+Query OK, 0 rows affected (0.11 sec)
+Query OK, 5 rows affected (0.01 sec)
+Records: 5  Duplicates: 0  Warnings: 0
```
Comment (TomShawn) [inline explain-aggregation.md:140]:
```suggestion
Query OK, 0 rows affected (0.11 sec)

Query OK, 5 rows affected (0.01 sec)
Records: 5  Duplicates: 0  Warnings: 0

```

File: explain-aggregation.md [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+
+{{< copyable "sql" >}}
+
+```sql
+ALTER TABLE t2 ADD INDEX (col1);
+EXPLAIN SELECT /*+ STREAM_AGG() */ col1, count(*) FROM t2 GROUP BY col1;
+```
+
+```sql
+Query OK, 0 rows affected (0.28 sec)
```
Comment (TomShawn) [inline explain-aggregation.md:163]:
```suggestion
Query OK, 0 rows affected (0.28 sec)

```

**Review by TomShawn — APPROVED** [review]
LGTM

**Pull request merged**